### PR TITLE
feat: multi gpu (one gpu at a time, single process) training

### PR DESCRIPTION
**Pull request by andre-mostly**
Change to allow large models to be run sharded across GPUs but single process -- to minimise code changes.

### CODE REVIEW

File: mostlyai/engine/_language/training.py [hunk 1]
```diff
@@ -293,13 +302,32 @@ def train(
...
             else (torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu"))
         )
+
+        if not with_dp:
+            if device.type == "cuda":
+                fsdp_plugin = FullyShardedDataParallelPlugin(
+                    state_dict_config=FullStateDictConfig(offload_to_cpu=True, rank0_only=False),
+                    optim_state_dict_config=FullOptimStateDictConfig(offload_to_cpu=True, rank0_only=False),
+                )
+                single_gpu_if_smaller_than = 7_000_000_000
```
[the remark this answers is not in the document]
Comment (andre-mostly) [reply mostlyai/engine/_language/training.py:319]: I like that better, will rename.

File: mostlyai/engine/_language/training.py [hunk 2]
```diff
@@ -681,9 +718,6 @@ def concat_prompt_and_response(x):
             current_lr = optimizer.param_groups[0][
                 "lr"
             ]  # currently assume that we have the same lr for all param groups
-            # only the scheduling for ReduceLROnPlateau is postponed until the metric becomes available
-            if not isinstance(lr_scheduler, torch.optim.lr_scheduler.ReduceLROnPlateau):
```
Comment (shuangwu5) [inline mostlyai/engine/_language/training.py:685]: Why removing the condition? I think it's better to leave them there, so that when someone wants to switch to a different LR scheduler, they don't have to worry about the training loop.
